### ISSUE TITLE
Multicam with individual Audio Role voice isolation breaks in FCPXML v1.11

### ISSUE THREAD

**Issue by randomeizer** [CommandPost/FCPCafe#314]
Apple Feedback Assistant ID: FB13500575
---

**DESCRIBE THE BUG:**
When I have a timeline with a Multicam clip that has multiple audio roles, and I select an individual role to apply Voice Isolation + Volume adjustments, the exported FCPXML in v1.11 will be output in this order (`volume` then `voiceIsolation`):

```xml
<mc-source angleID="VxamD/tAS4iO44cSAPFyuw" srcEnable="audio">
    <audio-role-source role="dialogue.Foo"/>
    <audio-role-source role="dialogue.Bar">
        <adjust-volume amount="12dB"/>
        <adjust-voiceIsolation amount="50"/>
    </audio-role-source>
</mc-source>
```

However, it should be in this order (`voiceIsolation`, then `volume`):

```xml
<mc-source angleID="VxamD/tAS4iO44cSAPFyuw" srcEnable="audio">
    <audio-role-source role="dialogue.Foo"/>
    <audio-role-source role="dialogue.Bar">
        <adjust-voiceIsolation amount="50"/>
        <adjust-volume amount="12dB"/>
    </audio-role-source>
</mc-source>
```

This only occurs in FCPXML 1.11. FCPXML 1.10 outputs correctly.

---

**TO REPRODUCE:**
1. Create a multicam clip with at least two different audio roles. Eg, make a Stereo track into Dual Mono, assign different roles to each.
2. Select only one of the roles.
3. Apply both volume and voice isolation adjustments to one of the roles.
4. Export the XML in v1.11
5. Try importing the same exported XML file back into Final Cut. Get this error message: ![Voice Isolation Import Error](https://github.com/CommandPost/FCPCafe/assets/3589315/4af571a3-73c8-47ee-8e86-51f2990036a7)

---

**EXPECTED BEHAVIOUR**
Exporting would be in the correct order, and importing would work correctly.

---

**SCREENSHOTS**
![Voice Isolation Bug - Inspector](https://github.com/CommandPost/FCPCafe/assets/3589315/2983819f-25cd-4311-ba7f-c5a825dd0917)
![Voice Isolation Import Error](https://github.com/CommandPost/FCPCafe/assets/3589315/4af571a3-73c8-47ee-8e86-51f2990036a7)

---

**SPECS**
- 2022 16-inch MacBook Pro (M2 Max, 32GB RAM, 1TB SSD)
- macOS Ventura 13.6.3 (22G436)
- Final Cut Pro 10.7.1

---

**ADDITIONAL COMMENTS**
**Workaround:** Export in FCPXML v1.10.

**Comment by randomeizer** [comment]
I've reported this to Apple but didn't get a feedback ID.

**Comment by latenitefilms** [comment]
Another example:

<img width="260" alt="invalid-fcpxml" src="https://github.com/CommandPost/FCPCafe/assets/22286696/fc097a10-028f-44f9-8b0b-47e4cedc3fb8">

**Comment by latenitefilms** [comment]
@randomeizer - Strangely, I'm having trouble manually re-creating this bug. Are you able to share a demonstration library?

**Comment by latenitefilms** [comment]
For any developers following along, here's my quick and dirty Swift-workaround:

```swift
import Foundation

//---------------------------------------------------------
// Correct FCPXML for a bug in Final Cut Pro 10.7.1 when
// using FCPXML v1.11.
//
// See: https://github.com/CommandPost/FCPCafe/issues/314
//---------------------------------------------------------
func correctFCPXML(xmlString: String) -> String {
    guard let xmlData = xmlString.data(using: .utf8) else {
        return xmlString
    }

    let parser = XMLParser(data: xmlData)
    let delegate = FCPXMLWorkaroundsParserDelegate()
    parser.delegate = delegate

    if parser.parse(), delegate.isVersionValid {
        NSLog("[Metaburner Pro Renderer] It's from Final Cut Pro v10.7.1 and FCPXML v1.11 so we'll 'fix it'.")
        return delegate.getCorrectedXML()
    } else {
        return xmlString
    }
}

//---------------------------------------------------------
// Parser Delegate:
//---------------------------------------------------------
class FCPXMLWorkaroundsParserDelegate: NSObject, XMLParserDelegate {
    private var correctedXML: String = ""
    private var currentElement: String = ""
    private var currentAudioRoleSource: String = ""
    var isVersionValid: Bool = false
    private var orderMap: [String: Int] = [
        "adjust-loudness":          1,
        "adjust-noiseReduction":    2,
        "adjust-humReduction":      3,
        "adjust-EQ":                4,
        "adjust-matchEQ":           5,
        "adjust-voiceIsolation":    6,
        "adjust-volume":            7,
        "adjust-panner":            8,
        "filter-audio":             9,
        "mute":                     10
    ]
    private var audioRoleSourceChildren: [String] = []

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        switch currentElement {
            case "audio-role-source":
                currentAudioRoleSource += string
            default:
                correctedXML += string
        }
    }

    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String : String] = [:]) {
        currentElement = elementName
        if elementName == "fcpxml" {
            if let version = attributeDict["version"], version == "1.11" {
                isVersionValid = true
            }
            correctedXML += "<\(elementName)"
            attributeDict.forEach { correctedXML += " \($0.key)=\"\($0.value)\"" }
            correctedXML += ">"
        } else if elementName == "audio-role-source" {
            currentAudioRoleSource = "<\(elementName)"
            attributeDict.forEach { currentAudioRoleSource += " \($0.key)=\"\($0.value)\"" }
            currentAudioRoleSource += ">"
        } else if !audioRoleSourceChildren.contains(elementName) {
            correctedXML += "<\(elementName)"
            attributeDict.forEach { correctedXML += " \($0.key)=\"\($0.value)\"" }
            correctedXML += ">"
        }
    }

    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        if elementName == "audio-role-source" {
            reorderAudioRoleSourceChildren()
            correctedXML += currentAudioRoleSource
            correctedXML += "</\(elementName)>"
        } else if !audioRoleSourceChildren.contains(elementName) {
            correctedXML += "</\(elementName)>"
        }
    }

    private func reorderAudioRoleSourceChildren() {
        let sortedChildren = audioRoleSourceChildren.sorted { orderMap[$0]! < orderMap[$1]! }
        sortedChildren.forEach { currentAudioRoleSource += $0 }
        audioRoleSourceChildren.removeAll()
    }

    func parser(_ parser: XMLParser, parseErrorOccurred parseError: Error) {
        NSLog("[Metaburner Pro Renderer] FCPXMLWorkarounds - Parse error: \(parseError.localizedDescription)")
    }

    func getCorrectedXML() -> String {
        return correctedXML
    }
}
```

**Comment by randomeizer** [comment]
Were you able to replicate? I didn't get time to export a library yesterday, sorry.

**Comment by randomeizer** [comment]
[Voice Isolation Bug.zip](https://github.com/CommandPost/FCPCafe/files/13814790/Voice.Isolation.Bug.zip)

Here is a library which contains a project which will export with the FCPXML bug in v1.11.

**Comment by latenitefilms** [comment]
Apple reports via Feedback Assistant:

> This issue was resolved in Final Cut Pro 10.8 release.
